### PR TITLE
App Submission: bleskomat-server

### PR DESCRIPTION
**Pull request by chill117**
# App Submission

### App name

Bleskomat Server

### Version

1.2.0

### One line description of the app

Connect a Bleskomat ATM to your Lightning node.

### Summary of the app

The Bleskomat ATM is the next generation Bitcoin Lightning ATM. This app will allow you to easily connect your Bleskomat ATM to the Lightning node on your Umbrel. A simple web interface is provided which you can use to manage and authorize one or more Bleskomat ATMs as well as view a list of recent payments handled by the server.

### Developer name

Bleskomat

### Developer website

https://www.bleskomat.com

### Source code link

https://github.com/samotari/bleskomat-server

### Support link

https://github.com/samotari/bleskomat-server/issues
https://t.me/bleskomat

### Requires

- [ ] Bitcoin Core
- [ ] Electrum server
- [x] LND

### 256x256 SVG icon

https://svgshare.com/i/d7L.svg

### Gallery images

Overview screen:
overview

General settings screen:
settings-general

Login credentials settings screen:
settings-login

Lightning configuration settings screen:
settings-lightning

Help screen:
help

Login screen:
login

### I have tested my app on:

- [ ] [Umbrel dev environment](https://github.com/getumbrel/umbrel-dev)
- [ ] [Umbrel OS on a Raspberry Pi 4](https://github.com/getumbrel/umbrel-os)
- [ ] [Custom Umbrel install on Linux](https://github.com/getumbrel/umbrel#-installation)

### CODE REVIEW

**Comment by mayankchhabra** [comment]
Thanks for the submission, @chill117! Bleskomat looks super cool! We'll review and test the app soon. 

Btw, test it out, do we need a real Bleskomat ATM machine to connect the app to? Or is there a way we can emulate the ATM, or even better, connect to any demo instance that you might be running?

**Comment by chill117** [comment]
> Thanks for the submission, @chill117! Bleskomat looks super cool! We'll review and test the app soon.

Thank you and glad that you like our project!

> Btw, test it out, do we need a real Bleskomat ATM machine to connect the app to? Or is there a way we can emulate the ATM, or even better, connect to any demo instance that you might be running?

It's possible to test the lnurl-withdraw flow without a Bleskomat ATM. Inside the docker image there is a node.js script which can generate signed LNURLs in the same way as the Bleskomat ATM. So once the bleskomat-server app is running in Umbrel, you can test as follows:

1) Access the bleskomat-server app in a browser - I think it should be available at [umbrel-dev.local:3333](http://umbrel-dev.local:3333).
2) Use the app password to login - this is the `APP_PASSWORD` environment variable which was passed to the bleskomat-server app.
3) Click the "New API Key" button to add a new API key
4) Run the following inside the bleskomat-server docker container to generate a signed LNURL:
  * `node ./scripts/generateSignedUrl.js 0.01 EUR | ./node_modules/.bin/lnurl encode`
  * The output will be a bech32-encoded string that you can copy/paste into a wallet app which supports lnurl-withdraw (LUD-03). A list of supporting apps can be found [here](https://github.com/fiatjaf/lnurl-rfc#lnurl-documents).
  * You could also encode the string as a QR code and scan it with a supporting mobile wallet.

Note that the bleskomat-server must be reachable by the mobile wallet app. I am not sure which mobile wallet apps support both lnurl-withdraw and onion addresses, so if you can use nginx or some other reverse proxy to expose the bleskomat-server publicly, that is more likely to succeed.

If you have any questions or change requests, I am happy to discuss. Thank you again for taking the time to look at this.

File: apps/bleskomat-server/data/entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+
+export BLESKOMAT_SERVER_ADMIN_PASSWORD=$(npm run --silent generate:adminPasswordHash -- "$APP_PASSWORD" | xargs)
```
Comment (lukechilds) [inline apps/bleskomat-server/data/entrypoint.sh:3]: Why use `xargs` here?

Comment (lukechilds) [reply apps/bleskomat-server/data/entrypoint.sh:3]: Also doesn't look like `$APP_PASSWORD` as actually being passed in from the compose file.

Comment (lukechilds) [reply apps/bleskomat-server/data/entrypoint.sh:3]: What about supporting this directly in your app? If your app accepted `BLESKOMAT_SERVER_ADMIN_PASSWORD_PLAINTEXT` and hashed it internally you wouldn't need the bash entrypoint wrapper, could handle errors safer (like empty password value) and it also wouldn't leak the plaintext password to other processes in the CLI flag for the `npm run` process.

This doesn't need to be a blocker for shipping on Umbrel, it's also ok how it's implemetned currently, but just a suggestion if you're open to changing it.

File: apps/bleskomat-server/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      DEBUG: bleskomat-server*,lnurl*
+      BLESKOMAT_SERVER_HOST: 0.0.0.0
+      BLESKOMAT_SERVER_PORT: $APP_BLESKOMAT_SERVER_PORT
+      BLESKOMAT_SERVER_URL: $APP_HIDDEN_SERVICE
+      BLESKOMAT_SERVER_ENDPOINT: /u
+      BLESKOMAT_SERVER_AUTH_API_KEYS: '[]'
+      BLESKOMAT_SERVER_LIGHTNING: '{"backend":"lnd","config":{"cert":"/lnd/tls.cert","protocol":"https","hostname":"$LND_IP:$LND_REST_PORT","macaroon":"/lnd/data/chain/bitcoin/$BITCOIN_NETWORK/admin.macaroon"}}'
+      BLESKOMAT_SERVER_STORE: '{"backend":"knex","config":{"client":"sqlite3","connection":{"filename":"/data/db.sqlite3"},"useNullAsDefault":true}}'
+      BLESKOMAT_SERVER_COINRATES_DEFAULTS_PROVIDER: coinbase
+      BLESKOMAT_SERVER_ADMIN_WEB: true
```
Comment (lukechilds) [inline apps/bleskomat-server/docker-compose.yml:25]: This causes a YAML parsing error, you need to quote the string:

```
services.web.environment.BLESKOMAT_SERVER_ADMIN_WEB contains true, which is an invalid type, it should be a string, number, or a null
```

```suggestion
      BLESKOMAT_SERVER_ADMIN_WEB: "true"
```

File: apps/bleskomat-server/data/entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
```
Comment (lukechilds) [inline apps/bleskomat-server/data/entrypoint.sh:1]: This file has not been set as executable so running the container fails with:

```
starting container process caused: exec: "/data/entrypoint.sh": permission denied: unknown
```

File: apps/bleskomat-server/data/entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+
+export BLESKOMAT_SERVER_ADMIN_PASSWORD=$(npm run --silent generate:adminPasswordHash -- "$APP_PASSWORD" | xargs)
+
+exec "/usr/src/app/docker-entrypoint.sh $@"
```
Comment (lukechilds) [inline apps/bleskomat-server/data/entrypoint.sh:5]: Does this file exist inside the container? It's failing for me with:

```
/data/entrypoint.sh: line 5: /usr/src/app/docker-entrypoint.sh : No such file or directory
```

File: apps/bleskomat-server/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,28 @@
+version: "3.7"
+
+services:
+  web:
+    image: bleskomat/bleskomat-server:1.2.0@sha256:1566bee49252c2aa431094f07fdb6ad1cb71eb0e916bfe7484f9d54e75b38bf1
```
Comment (lukechilds) [inline apps/bleskomat-server/docker-compose.yml:5]: This is pinning to the hash of the arm64 image. We need to pin to the multiarch manifest so the app works on both arm64 and amd64 Umbrel installs.

```suggestion
    image: bleskomat/bleskomat-server:1.2.0@sha256:278703d6ada3f038158bf3799dcefaa4d38cffff25e63e20f0bced23ef1ae5ca
```

For future reference you can get the multiarch manifest hash from `docker pull`:

```
$ docker pull bleskomat/bleskomat-server:1.2.0
1.2.0: Pulling from bleskomat/bleskomat-server
Digest: sha256:278703d6ada3f038158bf3799dcefaa4d38cffff25e63e20f0bced23ef1ae5ca
Status: Image is up to date for bleskomat/bleskomat-server:1.2.0
docker.io/bleskomat/bleskomat-server:1.2.0
```

File: apps/bleskomat-server/data/entrypoint.sh (hunk 4 above)
Comment (lukechilds) [reply apps/bleskomat-server/data/entrypoint.sh:5]: Ah it's cos you're quoting the whole lot so you're actually trying to execute "/usr/src/app/docker-entrypoint.sh " when `$@` is unset (notive the trailing space).

This should be what you want:

```suggestion
exec "/usr/src/app/docker-entrypoint.sh" $@
```

**Review by lukechilds — CHANGES_REQUESTED** [review]
After applying all the fixes mentioned in the review in my local branch I'm able to get the node process to start but it fails with:

```
web_1  | npm WARN tar ENOENT: no such file or directory, lstat '/usr/src/app/node_modules/.staging/knex-fb2d704c/lib/migrations'
web_1  | npm WARN tar ENOENT: no such file or directory, lstat '/usr/src/app/node_modules/.staging/knex-fb2d704c/lib/migrations'
web_1  | npm WARN tar ENOENT: no such file or directory, lstat '/usr/src/app/node_modules/.staging/knex-fb2d704c/lib/migrations'
web_1  | npm WARN tar ENOENT: no such file or directory, lstat '/usr/src/app/node_modules/.staging/knex-fb2d704c/lib/migrations'
web_1  | npm WARN tar ENOENT: no such file or directory, lstat '/usr/src/app/node_modules/.staging/knex-fb2d704c/types'
web_1  | npm WARN tar ENOENT: no such file or directory, lstat '/usr/src/app/node_modules/.staging/knex-fb2d704c/types'
web_1  | npm WARN tar ENOENT: no such file or directory, lstat '/usr/src/app/node_modules/.staging/knex-fb2d704c/types'
web_1  | npm ERR! code ENOENT
web_1  | npm ERR! syscall spawn git
web_1  | npm ERR! path git
web_1  | npm ERR! errno ENOENT
web_1  | npm ERR! enoent Error while executing:
web_1  | npm ERR! enoent undefined ls-remote -h -t ssh://git@github.com/samotari/open-sans.git
web_1  | npm ERR! enoent
web_1  | npm ERR! enoent
web_1  | npm ERR! enoent spawn git ENOENT
web_1  | npm ERR! enoent This is related to npm not being able to find a file.
web_1  | npm ERR! enoent
web_1  | npm WARN tar ENOENT: no such file or directory, lstat '/usr/src/app/node_modules/.staging/knex-fb2d704c/scripts/stress-test'
web_1  | npm WARN tar ENOENT: no such file or directory, open '/usr/src/app/node_modules/.staging/knex-fb2d704c/scripts/docker-compose.yml'
web_1  | npm WARN tar ENOENT: no such file or directory, lstat '/usr/src/app/node_modules/.staging/knex-fb2d704c/scripts/stress-test'
web_1  |
web_1  | npm ERR! A complete log of this run can be found in:
web_1  | npm ERR!     /home/node/.npm/_logs/2022-01-09T11_56_14_106Z-debug.log
```

Have you tested this and got it working on your end?

File: apps/bleskomat-server/docker-compose.yml (hunk 2 above)
Comment (lukechilds) [inline apps/bleskomat-server/docker-compose.yml:25]: You need to pass `$APP_PASSWORD` through from the compose configuration to the container:
```suggestion
      BLESKOMAT_SERVER_ADMIN_WEB: true
      APP_PASSWORD: $APP_PASSWORD
```

File: apps/bleskomat-server/data/entrypoint.sh (hunk 1 above)
Comment (chill117) [reply apps/bleskomat-server/data/entrypoint.sh:3]: > Why use `xargs` here?

xargs is used to trim white-space from around the password hash.

> Also doesn't look like $APP_PASSWORD as actually being passed in from the compose file.

I saw it used in other apps in Umbel, but didn't realize it was necessary to proxy its value via the docker compose file.

> What about supporting this directly in your app? If your app accepted BLESKOMAT_SERVER_ADMIN_PASSWORD_PLAINTEXT and hashed it internally you wouldn't need the bash entrypoint wrapper, could handle errors safer (like empty password value) and it also wouldn't leak the plaintext password to other processes in the CLI flag for the npm run process.

This sounds reasonable. I will add the new environment variable to the bleskomat-server project.

File: apps/bleskomat-server/docker-compose.yml (hunk 5 above)
Comment (chill117) [reply apps/bleskomat-server/docker-compose.yml:5]: Thanks for the tip. I'll be sure to use the multiarch sha256 hash in the future.

**Comment by chill117** [comment]
Thanks for taking the time to review. I'll review all of your suggestions and ping you here once the pull-request is ready for another look.

File: apps/bleskomat-server/docker-compose.yml (hunk 2 above)
Comment (chill117) [reply apps/bleskomat-server/docker-compose.yml]: > Have you tested this and got it working on your end?

I tried testing the new app with a local umbrel-dev instance. But I got stuck during the boot step of the umbrel-dev VM:
```
umbrel-dev: Unable to find image 'getumbrel/tor:v0.4.1.9' locally
```

**Comment by chill117** [comment]
@lukechilds I've made the requested changes and re-based with master to fix conflicts. Ready for another review.

**Review by lukechilds — CHANGES_REQUESTED** [review]
I get an error running the container. It looks look you're attempting to install dependencies at runtime but the Docker container doesn't have the required build tooling installed: https://github.com/samotari/bleskomat-server/blob/fd1b22032ff8a658a9f8ea33d5854244fd13213a/docker-entrypoint.sh#L28

```
web_1  | bleskomat-server@1.3.2 /usr/src/app
web_1  | `-- (empty)
web_1  |
web_1  | npm WARN deprecated node-pre-gyp@0.11.0: Please upgrade to @mapbox/node-pre-gyp: the non-scoped node-pre-gyp package is deprecated and only the @mapbox scoped package will recieve updates in the future
web_1  | npm WARN deprecated tar@2.2.2: This version of tar is no longer supported, and will not receive security updates. Please upgrade asap.
web_1  | npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
web_1  | npm WARN deprecated har-validator@5.1.5: this library is no longer supported
web_1  | npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
web_1  |
web_1  | > sqlite3@5.0.2 install /usr/src/app/node_modules/sqlite3
web_1  | > node-pre-gyp install --fallback-to-build
web_1  |
web_1  | node-pre-gyp WARN Using request for node-pre-gyp https download
web_1  | node-pre-gyp WARN Tried to download(403): https://mapbox-node-binary.s3.amazonaws.com/sqlite3/v5.0.2/napi-v3-linux-arm64.tar.gz
web_1  | node-pre-gyp WARN Pre-built binaries not found for sqlite3@5.0.2 and node@12.22.9 (node-v72 ABI, glibc) (falling back to source compile with node-gyp)
web_1  | gyp ERR! find Python
web_1  | gyp ERR! find Python Python is not set from command line or npm configuration
web_1  | gyp ERR! find Python Python is not set from environment variable PYTHON
web_1  | gyp ERR! find Python checking if "python" can be used
web_1  | gyp ERR! find Python - "python" is not in PATH or produced an error
web_1  | gyp ERR! find Python checking if "python2" can be used
web_1  | gyp ERR! find Python - "python2" is not in PATH or produced an error
web_1  | gyp ERR! find Python checking if "python3" can be used
web_1  | gyp ERR! find Python - "python3" is not in PATH or produced an error
web_1  | gyp ERR! find Python
web_1  | gyp ERR! find Python **********************************************************
web_1  | gyp ERR! find Python You need to install the latest version of Python.
web_1  | gyp ERR! find Python Node-gyp should be able to find and use Python. If not,
web_1  | gyp ERR! find Python you can try one of the following options:
web_1  | gyp ERR! find Python - Use the switch --python="/path/to/pythonexecutable"
web_1  | gyp ERR! find Python   (accepted by both node-gyp and npm)
web_1  | gyp ERR! find Python - Set the environment variable PYTHON
web_1  | gyp ERR! find Python - Set the npm configuration variable python:
web_1  | gyp ERR! find Python   npm config set python "/path/to/pythonexecutable"
web_1  | gyp ERR! find Python For more information consult the documentation at:
web_1  | gyp ERR! find Python https://github.com/nodejs/node-gyp#installation
web_1  | gyp ERR! find Python **********************************************************
web_1  | gyp ERR! find Python
web_1  | gyp ERR! configure error
web_1  | gyp ERR! stack Error: Could not find any Python installation to use
web_1  | gyp ERR! stack     at PythonFinder.fail (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:307:47)
web_1  | gyp ERR! stack     at PythonFinder.runChecks (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:136:21)
web_1  | gyp ERR! stack     at PythonFinder.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:179:16)
web_1  | gyp ERR! stack     at PythonFinder.execFileCallback (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/find-python.js:271:16)
web_1  | gyp ERR! stack     at exithandler (child_process.js:315:5)
web_1  | gyp ERR! stack     at ChildProcess.errorhandler (child_process.js:327:5)
web_1  | gyp ERR! stack     at ChildProcess.emit (events.js:314:20)
web_1  | gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:274:12)
web_1  | gyp ERR! stack     at onErrorNT (internal/child_process.js:470:16)
web_1  | gyp ERR! stack     at processTicksAndRejections (internal/process/task_queues.js:84:21)
web_1  | gyp ERR! System Linux 5.10.52-v8+
web_1  | gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "configure" "--fallback-to-build" "--module=/usr/src/app/node_modules/sqlite3/lib/binding/napi-v3-linux-arm64/node_sqlite3.node" "--module_name=node_sqlite3" "--module_path=/u
sr/src/app/node_modules/sqlite3/lib/binding/napi-v3-linux-arm64" "--napi_version=8" "--node_abi_napi=napi" "--napi_build_version=3" "--node_napi_label=napi-v3"
web_1  | gyp ERR! cwd /usr/src/app/node_modules/sqlite3
web_1  | gyp ERR! node -v v12.22.9
web_1  | gyp ERR! node-gyp -v v5.1.0
web_1  | gyp ERR! not ok
web_1  | node-pre-gyp ERR! build error
web_1  | node-pre-gyp ERR! stack Error: Failed to execute '/usr/local/bin/node /usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --module=/usr/src/app/node_modules/sqlite3/lib/binding/napi-v3-linux-arm64/node_sqlite3.node --module_name=node_sqli
te3 --module_path=/usr/src/app/node_modules/sqlite3/lib/binding/napi-v3-linux-arm64 --napi_version=8 --node_abi_napi=napi --napi_build_version=3 --node_napi_label=napi-v3' (1)
web_1  | node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (/usr/src/app/node_modules/node-pre-gyp/lib/util/compile.js:83:29)
web_1  | node-pre-gyp ERR! stack     at ChildProcess.emit (events.js:314:20)
web_1  | node-pre-gyp ERR! stack     at maybeClose (internal/child_process.js:1022:16)
web_1  | node-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:287:5)
web_1  | node-pre-gyp ERR! System Linux 5.10.52-v8+
web_1  | node-pre-gyp ERR! command "/usr/local/bin/node" "/usr/src/app/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build"
web_1  | node-pre-gyp ERR! cwd /usr/src/app/node_modules/sqlite3
web_1  | node-pre-gyp ERR! node -v v12.22.9
web_1  | node-pre-gyp ERR! node-pre-gyp -v v0.11.0
web_1  | node-pre-gyp ERR! not ok
web_1  | Failed to execute '/usr/local/bin/node /usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --module=/usr/src/app/node_modules/sqlite3/lib/binding/napi-v3-linux-arm64/node_sqlite3.node --module_name=node_sqlite3 --module_path=/usr/src/app/
node_modules/sqlite3/lib/binding/napi-v3-linux-arm64 --napi_version=8 --node_abi_napi=napi --napi_build_version=3 --node_napi_label=napi-v3' (1)
web_1  | npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@2.3.2 (node_modules/fsevents):
web_1  | npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"arm64"})
web_1  |
web_1  | npm ERR! code ELIFECYCLE
web_1  | npm ERR! errno 1
web_1  | npm ERR! sqlite3@5.0.2 install: `node-pre-gyp install --fallback-to-build`
web_1  | npm ERR! Exit status 1
```

Generally running a Docker container should be deterministic and reproducible. They should have all dependencies setup as part of building the the image. If you want to reduce the image size by not distributing unused deps it's common to do this by building multiple versions of the image with different deps that have different tags. So consumers can just pull the tag with the deps they need.

If you need to install deps at runtime, which I would advise, you should add all potential required build deps to the base image so all the deps can definitely be built correctly. You might find this increases the image size more than just distributing all potential deps and not shipping build deps in the resulting image.

**Comment by chill117** [comment]
@lukechilds Ready for another review. What was changed:
* Removed the custom entrypoint file in the bleskomat-server docker image - so dependencies will no longer be installed at run-time
* The docker image includes pre-installed data store dependencies for postgres
* The bleskomat-server umbrel app will now use postgres as its data store

**Comment by lukechilds** [comment]
This is now working for me, awesome work! 🚀

There is one issue I spotted. The node process doesn't seem to wait for the DB to become available. So on the first start after installation, the DB takes a little longer to initialise and the connection from node fails. Once the DB has initialised, all future DB interactions fail too. You can see this below:

```
Attaching to bleskomat-server_db_1, bleskomat-server_web_1                                                                                                                                                                                                                                    [31/1848]
db_1   | chmod: changing permissions of '/var/run/postgresql': Operation not permitted
db_1   | The files belonging to this database system will be owned by user "postgres".
db_1   | This user must also own the server process.
db_1   |
db_1   | The database cluster will be initialized with locale "en_US.utf8".
db_1   | The default database encoding has accordingly been set to "UTF8".
db_1   | The default text search configuration will be set to "english".
db_1   |
db_1   | Data page checksums are disabled.
db_1   |
db_1   | fixing permissions on existing directory /var/lib/postgresql/data ... ok
db_1   | creating subdirectories ... ok
db_1   | selecting default max_connections ... 100
db_1   | selecting default shared_buffers ... 128MB
db_1   | selecting default timezone ... Etc/UTC
db_1   | selecting dynamic shared memory implementation ... posix
db_1   | creating configuration files ... ok
db_1   | running bootstrap script ... ok
web_1  |
web_1  | > bleskomat-server@1.3.3 start /usr/src/app
web_1  | > node ./index.js
web_1  |
db_1   | performing post-bootstrap initialization ... ok
db_1   | syncing data to disk ... ok
db_1   |
db_1   | WARNING: enabling "trust" authentication for local connections
db_1   | You can change this by editing pg_hba.conf or using the option -A, or
db_1   | --auth-local and --auth-host, the next time you run initdb.
db_1   |
db_1   | Success. You can now start the database server using:
db_1   |
db_1   |     pg_ctl -D /var/lib/postgresql/data -l logfile start
db_1   |
web_1  | 2022-02-09T06:43:18.473Z lnurl:info Creating web server...
db_1   | waiting for server to start....2022-02-09 06:43:18.478 UTC [41] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
db_1   | 2022-02-09 06:43:18.555 UTC [42] LOG:  database system was shut down at 2022-02-09 06:43:17 UTC
db_1   | 2022-02-09 06:43:18.580 UTC [41] LOG:  database system is ready to accept connections
db_1   |  done
db_1   | server started
web_1  | 2022-02-09T06:43:18.862Z lnurl:store:memory:info Store initialized and ready for use
web_1  | 2022-02-09T06:43:18.864Z lnurl:info Creating web server...
web_1  | 2022-02-09T06:43:19.575Z lnurl:info Web server listening at http://0.0.0.0:3333/
web_1  | 2022-02-09T06:43:19.609Z lnurl:store:knex:error Database migrations failed: Error: connect ECONNREFUSED 10.21.21.77:5432
web_1  |     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1159:16) {
web_1  |   errno: -111,
web_1  |   code: 'ECONNREFUSED',
web_1  |   syscall: 'connect',
web_1  |   address: '10.21.21.77',
web_1  |   port: 5432
web_1  | }
web_1  | (node:20) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 10.21.21.77:5432
web_1  |     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1159:16)
web_1  | (Use `node --trace-warnings ...` to show where the warning was created)
web_1  | (node:20) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejecti
on, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 5)
web_1  | (node:20) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
db_1   | CREATE DATABASE
db_1   |
db_1   |
db_1   | /usr/local/bin/docker-entrypoint.sh: ignoring /docker-entrypoint-initdb.d/*
db_1   |
db_1   | 2022-02-09 06:43:19.796 UTC [41] LOG:  received fast shutdown request
db_1   | waiting for server to shut down....2022-02-09 06:43:19.807 UTC [41] LOG:  aborting any active transactions
db_1   | 2022-02-09 06:43:19.817 UTC [41] LOG:  worker process: logical replication launcher (PID 48) exited with exit code 1
db_1   | 2022-02-09 06:43:19.819 UTC [43] LOG:  shutting down
db_1   | 2022-02-09 06:43:19.882 UTC [41] LOG:  database system is shut down
db_1   |  done
db_1   | server stopped
db_1   |
db_1   | PostgreSQL init process complete; ready for start up.
db_1   |
db_1   | 2022-02-09 06:43:19.947 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
db_1   | 2022-02-09 06:43:19.949 UTC [1] LOG:  listening on IPv6 address "::", port 5432
db_1   | 2022-02-09 06:43:19.967 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
db_1   | 2022-02-09 06:43:20.034 UTC [69] LOG:  database system was shut down at 2022-02-09 06:43:19 UTC
db_1   | 2022-02-09 06:43:20.057 UTC [1] LOG:  database system is ready to accept connections
web_1  | 2022-02-09T06:43:28.912Z lnurl:request GET /admin/login
web_1  | 2022-02-09T06:43:29.569Z lnurl:request GET /css/all.css?v=1.3.3
web_1  | 2022-02-09T06:43:29.575Z lnurl:request GET /js/all.js?v=1.3.3
web_1  | 2022-02-09T06:43:29.601Z lnurl:request GET /images/bleskomat-server-logo.png
web_1  | 2022-02-09T06:43:29.613Z lnurl:request GET /fonts/OpenSans/OpenSans-Regular.woff2
web_1  | 2022-02-09T06:43:29.616Z lnurl:request GET /fonts/OpenSans/OpenSans-Light.woff2
web_1  | 2022-02-09T06:43:29.626Z lnurl:request GET /fonts/OpenSans/OpenSans-Bold.woff2
web_1  | 2022-02-09T06:43:29.628Z lnurl:request GET /fonts/OpenSans/OpenSans-LightItalic.woff2
web_1  | 2022-02-09T06:43:29.669Z lnurl:request GET /fonts/OpenSans/OpenSans-Regular.woff2
web_1  | 2022-02-09T06:43:29.678Z lnurl:request GET /fonts/OpenSans/OpenSans-Bold.woff2
web_1  | 2022-02-09T06:43:29.682Z lnurl:request GET /fonts/OpenSans/OpenSans-LightItalic.woff2
web_1  | 2022-02-09T06:43:29.686Z lnurl:request GET /fonts/OpenSans/OpenSans-Light.woff2
web_1  | 2022-02-09T06:43:29.732Z lnurl:request GET /images/favicon/favicon-196x196.png
web_1  | 2022-02-09T06:43:29.735Z lnurl:request GET /images/favicon/favicon-32x32.png
web_1  | 2022-02-09T06:43:30.000Z lnurl:request GET /admin/login
web_1  | 2022-02-09T06:43:30.002Z lnurl:request GET /admin/login
web_1  | 2022-02-09T06:43:48.779Z lnurl:request POST /admin/login
web_1  | 2022-02-09T06:43:49.118Z lnurl:request GET /admin
web_1  | 2022-02-09T06:43:49.153Z lnurl:request GET /admin/overview
web_1  | (node:20) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 10.21.21.77:5432
web_1  |     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1159:16)
web_1  | (node:20) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejecti
on, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 6)
```

Simply restarting the node container once the DB is up resolves the issue and everything works correctly:

```
bleskomat-server_web_1 exited with code 0
web_1  |
web_1  | > bleskomat-server@1.3.3 start /usr/src/app
web_1  | > node ./index.js
web_1  |
web_1  | 2022-02-09T06:44:20.817Z lnurl:info Creating web server...
web_1  | 2022-02-09T06:44:20.981Z lnurl:store:memory:info Store initialized and ready for use
web_1  | 2022-02-09T06:44:20.981Z lnurl:info Creating web server...
web_1  | 2022-02-09T06:44:21.190Z lnurl:info Web server listening at http://0.0.0.0:3333/
web_1  | 2022-02-09T06:44:21.643Z lnurl:store:knex:info Database migrations completed
web_1  | 2022-02-09T06:44:21.645Z lnurl:store:knex:info Store initialized and ready for use
```

I think there should be some additional logic in the node process to handle the DB not being available gracefully and wait for it to come online, then connect.

**Comment by chill117** [comment]
> Simply restarting the node container once the DB is up resolves the issue and everything works correctly:
> 
> ```
> bleskomat-server_web_1 exited with code 0
> web_1  |
> web_1  | > bleskomat-server@1.3.3 start /usr/src/app
> web_1  | > node ./index.js
> web_1  |
> web_1  | 2022-02-09T06:44:20.817Z lnurl:info Creating web server...
> web_1  | 2022-02-09T06:44:20.981Z lnurl:store:memory:info Store initialized and ready for use
> web_1  | 2022-02-09T06:44:20.981Z lnurl:info Creating web server...
> web_1  | 2022-02-09T06:44:21.190Z lnurl:info Web server listening at http://0.0.0.0:3333/
> web_1  | 2022-02-09T06:44:21.643Z lnurl:store:knex:info Database migrations completed
> web_1  | 2022-02-09T06:44:21.645Z lnurl:store:knex:info Store initialized and ready for use
> ```
> 
> I think there should be some additional logic in the node process to handle the DB not being available gracefully and wait for it to come online, then connect.

I added `"depends_on": db` to the docker-compose file. This should force compose to wait until the db container is up before starting the web container. This should fix the ECONNREFUSED error.

**Comment by lukechilds** [comment]
`depends_on` just waits for the process to start, it doesn't know anything about what the process is doing and when it's ready to accept conections. e.g if `db` process fails to start then `web` won't start. If `db` process starts then `web` will start. It has no idea if the `db` process has fully intialised and is ready to accept connections.

https://docs.docker.com/compose/startup-order/

**Comment by chill117** [comment]
@lukechilds Added the suggested database connection behavior to the upstream dependency (lnurl-node). It will now wait 30 seconds (default) for a database connection before failing.

**Comment by lukechilds** [comment]
Sorry for the delay @chill117, tested and working great!

Quick question, is there a way to verify the LND connection is definitely working? I see you're passing in the LND connection details, however the "Lightning Configuration" page shows blank inputs as if there is no connection.

Screenshot 2022-02-25 at 18 13 08

**Review by lukechilds — APPROVED** [review]

**Comment by lukechilds** [comment]
Thanks @chill117!

**Comment by mayankchhabra** [comment]
Really nice work, @chill117! Congrats on the [launch](https://twitter.com/getumbrel/status/1498665787523231744) 🚀

Here are the gallery images and app icon for [Bleskomat Server](https://github.com/getumbrel/umbrel-apps-gallery/tree/master/bleskomat-server). If you ever want to change/update them, feel free to directly open a PR at [getumbrel/umbrel-apps-gallery](https://github.com/getumbrel/umbrel-apps-gallery). Cheers!

**Pull request merged**